Problem compiling unicode

I tried to compile the following unicode and the compiler failed.

NOTE: Sure this is really an issue but wanted to document immediatly

    E       UnicodeEncodeError: 'ascii' codec can't encode character u'\u212b' in position 0: ordinal not in range(128)  pywbem/mof_compiler.py:1798: UnicodeEncodeError


        str = u'instance of CIM_Foo_sub2 as $foosub22 {' \
              u'InstanceID = "CIM_Foo_unicode";\n' \
              u'cimfoo_sub2 = \u212b \u0420 \u043e \u0441 \u0441' \
              u' \u0438 \u044f \u00e0;};'

generally same in python 3:

      raise MOFParseError(parser_token=p, msg=msg)

    E pywbem.mof_compiler.MOFParseError: MOFParseError:
    E None:2:14 msg=Illegal character 'Å'
    E ['instance of CIM_Foo_sub2 as $foosub22 {InstanceID = "CIM_Foo_unicode";', 'cimfoo_sub2 = Å   Р о с с и я à;};', ' ^^^^^^^^^^^^^^^^^^']

pywbem/mof_compiler.py:460: MOFParseError

--- a/docs/changes.rst
+++ b/docs/changes.rst
@@ -87,6 +87,12 @@ Released: not yet
 * Fix issue causing pywbem_mock/_wbemconnection_mock.py display_repository()
   to display as bytes in Python 3.  See issue # 1276
 
+* Fixed the support for Unicode escape sequences in the MOF compiler. It
+  supported only lower case `\x1234` but not upper case `\X1234`.
+  Also, it tolerated an invalid `\x` escape sequence, when DSP0004 requires
+  1..4 hex characters to be present.
+  See issue #1278.
+
 **Enhancements:**
 
 * Extend pywbem MOF compiler to search for dependent classes including:

--- a/pywbem/mof_compiler.py
+++ b/pywbem/mof_compiler.py
@@ -287,7 +287,7 @@ def t_decimalValue(t):
 
 
 simpleEscape = r"""[bfnrt'"\\]"""
-hexEscape = r'x[0-9a-fA-F]{1,4}'
+hexEscape = r'[xX][0-9a-fA-F]{1,4}'
 escapeSequence = r'[\\]((%s)|(%s))' % (simpleEscape, hexEscape)
 cChar = r"[^'\\\n\r]|(%s)" % escapeSequence
 sChar = r'[^"\\\n\r]|(%s)' % escapeSequence
@@ -750,7 +750,7 @@ def p_pragmaName(p):
 
 def p_pragmaParameter(p):
     """pragmaParameter : stringValue"""
-    p[0] = _fixStringValue(p[1])
+    p[0] = _fixStringValue(p[1], p)
 
 
 def p_classDeclaration(p):
@@ -1305,7 +1305,7 @@ def p_constantValueList(p):
         p[0] = p[1] + [p[3]]
 
 
-def _fixStringValue(s):
+def _fixStringValue(s, p):
     """Clean up string value including special characters, etc."""
 
     # pylint: disable=too-many-branches
@@ -1352,7 +1352,13 @@ def _fixStringValue(s):
                 else:
                     hexc |= ord(c) - ord('A') + 0XA
                 j += 1
-            rv += chr(hexc)
+            if j == 0:
+                # DSP0004 requires 1..4 hex chars - we have 0
+                raise MOFParseError(
+                    parser_token=p,
+                    msg="Unicode escape sequence (e.g. '\\x12AB') requires "
+                        "at least one hex character")
+            rv += six.unichr(hexc)
             i += j - 1
 
         esc = False
@@ -1365,9 +1371,9 @@ def p_stringValueList(p):
                        | stringValueList stringValue
                        """
     if len(p) == 2:
-        p[0] = _fixStringValue(p[1])
+        p[0] = _fixStringValue(p[1], p)
     else:
-        p[0] = p[1] + _fixStringValue(p[2])
+        p[0] = p[1] + _fixStringValue(p[2], p)
 
 
 def p_constantValue(p):

--- a/testsuite/test_mof_compiler.py
+++ b/testsuite/test_mof_compiler.py
@@ -8,7 +8,7 @@
 from __future__ import print_function, absolute_import
 
 import os
-import unittest
+import unittest2 as unittest  # we use assertRaises(exc) introduced in py27
 import six
 from ply import lex
 try:
@@ -82,6 +82,99 @@ class MOFTest(unittest.TestCase):
         if self.partial_schema_file:
             if os.path.exists(self.partial_schema_file):
                 os.remove(self.partial_schema_file)
+
+
+class TestInstancesUnicode(MOFTest):
+    """
+    Test compile of an instance with Unicode characters in string properties
+    """
+
+    def test_instance_unicode_literal(self):
+        """
+        Test compile of string property with literal Unicode characters
+        """
+
+        repo = self.mofcomp.handle
+
+        class_mof = \
+            'class PyWBEM_TestUnicode {\n' \
+            '    string uprop;\n' \
+            '};'
+        self.mofcomp.compile_string(class_mof, NAME_SPACE)
+
+        inst_mof = \
+            u'instance of PyWBEM_TestUnicode {\n' \
+            u'    uprop = "\u212b \u0420 \u043e \u0441 \u0441"\n' \
+            u'            "\u0438 \u044f \u00e0";\n' \
+            u'};'
+        exp_uprop = u"\u212b \u0420 \u043e \u0441 \u0441" \
+                    u"\u0438 \u044f \u00e0"
+
+        self.mofcomp.compile_string(inst_mof, NAME_SPACE)
+
+        for inst in repo.instances[NAME_SPACE]:
+            if inst.classname == 'PyWBEM_TestUnicode':
+                break
+        else:
+            inst = None
+        assert inst is not None
+
+        uprop = inst.properties['uprop'].value
+
+        self.assertEqual(uprop, exp_uprop)
+
+    def test_instance_unicode_escape(self):
+        """
+        Test compile of string property with MOF escaped Unicode characters
+        """
+
+        repo = self.mofcomp.handle
+
+        class_mof = \
+            'class PyWBEM_TestUnicode {\n' \
+            '    string uprop;\n' \
+            '};'
+        self.mofcomp.compile_string(class_mof, NAME_SPACE)
+
+        inst_mof = \
+            u'instance of PyWBEM_TestUnicode {\n' \
+            u'    uprop = "\\X212B \\X0420 \\x043E \\x0441 \\x0441 ' \
+            u'\\x0438 \\x044f \\x00e0";\n' \
+            u'};'
+        exp_uprop = u"\u212b \u0420 \u043e \u0441 \u0441 " \
+                    u"\u0438 \u044f \u00e0"
+
+        self.mofcomp.compile_string(inst_mof, NAME_SPACE)
+
+        for inst in repo.instances[NAME_SPACE]:
+            if inst.classname == 'PyWBEM_TestUnicode':
+                break
+        else:
+            inst = None
+        assert inst is not None
+
+        uprop = inst.properties['uprop'].value
+
+        self.assertEqual(uprop, exp_uprop)
+
+    def test_instance_unicode_escape_fail1(self):
+        """
+        Test compile of string property with Unicode escape with no hex chars
+        """
+
+        class_mof = \
+            'class PyWBEM_TestUnicode {\n' \
+            '    string uprop;\n' \
+            '};'
+        self.mofcomp.compile_string(class_mof, NAME_SPACE)
+
+        inst_mof = \
+            u'instance of PyWBEM_TestUnicode {\n' \
+            u'    uprop = "\\xzzzz";\n' \
+            u'};'
+
+        with self.assertRaises(MOFParseError):
+            self.mofcomp.compile_string(inst_mof, NAME_SPACE)
 
 
 class TestFlavors(MOFTest):

